pkg/generators/sentinel/statefulset.go

```diff
@@ -0,0 +1,139 @@
+package sentinel
+
+import (
+	"fmt"
+	"strings"
+
+	saasv1alpha1 "github.com/3scale/saas-operator/api/v1alpha1"
+	"github.com/3scale/saas-operator/pkg/basereconciler"
+	"github.com/3scale/saas-operator/pkg/generators/common_blocks/pod"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	healthCommand string = fmt.Sprintf("redis-cli -p %d PING", saasv1alpha1.SentinelPort)
+)
+
+// StatefulSet returns a basereconciler.GeneratorFunction function that will return
+// a StatefulSet resource when called
+func (gen *Generator) StatefulSet() basereconciler.GeneratorFunction {
+
+	return func() client.Object {
+		return &appsv1.StatefulSet{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "StatefulSet",
+				APIVersion: appsv1.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      gen.GetComponent(),
+				Namespace: gen.Namespace,
+				Labels:    gen.GetLabels(),
+			},
+			Spec: appsv1.StatefulSetSpec{
+				PodManagementPolicy: appsv1.ParallelPodManagement,
+				Replicas:            gen.Spec.Replicas,
+				Selector:            gen.Selector(),
+				ServiceName:         gen.GetComponent() + "-headless",
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: gen.LabelsWithSelector(),
+					},
+					Spec: corev1.PodSpec{
+						Affinity:                     pod.Affinity(gen.Selector().MatchLabels, gen.Spec.NodeAffinity),
+						AutomountServiceAccountToken: pointer.Bool(false),
+						ImagePullSecrets: func() []corev1.LocalObjectReference {
+							if gen.Spec.Image.PullSecretName != nil {
+								return []corev1.LocalObjectReference{{Name: *gen.Spec.Image.PullSecretName}}
+							}
+							return nil
+						}(),
+						Containers: []corev1.Container{
+							{
+								Command:         []string{"redis-server", "/redis/sentinel.conf", "--sentinel"},
+								Image:           fmt.Sprintf("%s:%s", *gen.Spec.Image.Name, *gen.Spec.Image.Tag),
+								ImagePullPolicy: *gen.Spec.Image.PullPolicy,
+								LivenessProbe: &corev1.Probe{
+									Handler: corev1.Handler{Exec: &corev1.ExecAction{
+										Command: strings.Split(healthCommand, " ")}},
+									FailureThreshold:    3,
```

Comment:
Shouldnt these liveness/readiness healthchecks be configurable at CRD level with sane default values?

Reply:
This is much a work in progress, just a first PR, so I didn't add it because I still need to test it in a real cluster. Furthermore, adding it now would cause me to suffer an unpleasant rebase experience which I'd rather avoid, so best leave it to a different PR :sweat_smile:

Reply:
hahaha fair enough!